src/Uno.UI.Runtime.Skia.WebAssembly.Browser/ts/Runtime/BrowserPointerInputSource.ts

```diff
@@ -81,6 +83,59 @@
 			element.addEventListener("wheel", this.onPointerEventReceived.bind(this), { capture: true, passive: false });
 		}
 
+
+		// Retrieve and cache the native element host reference.
+		// Refreshes if the node was detached or replaced (e.g., during hot reload).
+		private getNativeElementHostCached(): HTMLElement | null {
+			if (this._nativeElementHost === null || this._nativeElementHost.isConnected === false) {
+				this._nativeElementHost = document.getElementById("uno-native-element-host") as HTMLElement | null;
+			}
+			return this._nativeElementHost;
+		}
+
+
+		// Returns true if the event originated from within the native host subtree.
+		// Traverses regular DOM first, then crosses Shadow DOM boundaries when required.
+		// Uses identity comparisons only; avoids selector matching, allocations, and redundant lookups.
+		private isEventFromNativeElementHost(eventTarget: EventTarget | null) {
+			const hostElement = this.getNativeElementHostCached();
+			if (hostElement === null) {
+				return false; // No host exists; nothing to filter.
+			}
+
+			let currentNode = eventTarget as Node | null;
+
+			while (currentNode !== null) {
+				// Fast identity comparison.
+				if (currentNode === hostElement) {
+					return true;
+				}
+
+				// Normal DOM climb first (fastest path)
+				const parent = (currentNode as any).parentNode as Node | null;
+				if (parent) {
+					currentNode = parent;
+					continue;
+				}
+
+				// Only if parentNode is null, check for a shadow boundary.
+				const rootNode: any = typeof (currentNode as any).getRootNode === "function"
+					? (currentNode as any).getRootNode()
+					: null;
```

Comment:
According to the support table, [the `getRootNode` is supported by all modern browser](https://caniuse.com/mdn-api_node_getrootnode), IMO we can safely use it without performing check:


```suggestion
				const rootNode = currentNode.getRootNode();
```